### DISCUSSION TITLE
Retrieve the current being deployed container ID

### DISCUSSION THREAD

**Discussion by louis-l** [caprover/caprover#2013] [Q&A]
Hi,

Im trying to achieve zero-downtime deployment in my Laravel app and after months of trying different approaches, Im able to get the 502 downtime down to second, but still have not got it working (zero downtime). 

For the context, Im using https://github.com/jackbrycesmith/laravel-caprover-template and I have moved everything from `entrypoint.sh` to `Dockerfile` (followed this [discussion](https://github.com/jackbrycesmith/laravel-caprover-template/issues/1)). There are a few stuff remain in the `entrypoint.sh` that cant be moved (here is my current file):

```shell
#!/bin/sh

php artisan config:cache
php artisan route:cache
php artisan event:cache

php artisan migrate --no-interaction --force

supervisord -c $LARAVEL_PATH/.deploy/production/config/supervisor.conf
```

The problem is: if the new version/deployment does not have migration to run, the downtime is almost zero (which is great). Otherwise, the downtime will equal to the time it takes to run the migration.

I end up with putting `RUN sleep 10` into my `Dockerfile` as a leeway but that is not a great solution.

So my next attempt is to write a script and put that at the end of `Dockerfile` (or after running `entrypoint.sh`) to manually and constantly check the current being deployed container is up and running:

```shell
if [ "$(docker inspect --format='{{json .State.Health.Status}}' <container_name_or_id>)" = '"healthy"' ]; then
    echo "Container is healthy"
else
    echo "Container is unhealthy"
fi
```

The question is: how to get `<container_name_or_id>` var?

I asked ChatGPT and got this answer but not sure if there is better approach?

> If you really need to access the container ID within the Dockerfile (which is not typical), you could technically write a script to extract the container ID at runtime and write it to a file during container startup. However, this approach is not recommended as it goes against the typical Docker container workflow and introduces unnecessary complexity.

P/S: or is there a way to fire local request to the app URL to check if its 200 OK?

Cheers,

**Comment by githubsaturn** [comment]
You can just use a healthcheck command in your Dockerfile:

https://github.com/caprover/caprover/discussions/1287#discussioncomment-6486935

**Reply by louis-l** [reply]
I did try Docker healthcheck command ages ago and for some reason it did not work.
I remember the healthcheck run ok but I still saw 502 page when refreshing the app. Eventually gave up because dont know the best way to debug that.  

Let me give it another try and will update here.

**Comment by thgh** [comment]
I think you are looking for https://github.com/caprover/caprover/issues/830#issuecomment-690764743

**Reply by louis-l** [reply]
My app does not have persistent data, so I believe that is the default order in Caprover (could be wrong 🤷‍♂️)

**Comment by louis-l** [comment]
I think I have answered my own question.
I believe the issue here is that the new app is always up and running, so using docker healthcheck does not help in this case, because it will always return 200 OK from the new app being built.
Close this for now. 
Thank you all for your time.

**Comment by louis-l** [comment]
@githubsaturn 

Do you know the Dockerfile `HEALTHCHECK` command runs in server shell or the container shell?

Im testing some ideas, but when I ssh to the server and run `curl localhost` I got this Caprover 404 error page:

![Screenshot 2025-01-26 at 9 52 07 am](https://github.com/user-attachments/assets/6d9e5c7d-d51b-4818-9a7a-4f032454a418)

Same thing happens when I access my Laravel container shell and run `curl localhost` (same response)

**Reply by githubsaturn** [reply]
HEALTHCHECK runs inside the container itself.
